GATT service builder

alexmoon opened the pull request:
This provides an alternate way to create GATT services. Instead of the proc macro, it has a builder interface to allow full control of the characteristics and descriptors that are created for the service. This is necessary for services like HID over GATT which require additional descriptors to be attached to the characteristics or for services that need to require security for access.

Because the Softdevice service creation APIs are order-dependent, this requires an exclusive reference to the Softdevice to be safe. Therefore, I've changed `Softdevice::enable` to return an exclusive reference that can be used during initialization. That reference is then downgraded (by the compiler) to a shared reference once `Softdevice::run` is called.

alexmoon commented:
@Dirbaio,  @bobmcwhirter, @lulf: I'm pretty happy with the builder API this PR provides. I've used it to create several services now and it works well.  The only concern I have before merging is the change to `Softdevice::enable` to return a `&mut` reference instead of a shared reference.

This is required because the builder API is only sound if it can guarantee no other calls to the underlying softdevice APIs are made while a service is being built. However, it is technically a breaking change because the `&mut` reference is not `Copy` so you either need to reborrow it to a `&` reference or move it in order to do things like pass it into a closure.

I could easily make a `Softdevice::enable_mut` version instead to maintain backwards compatibility, but that complicates the API slightly. I'd like your input before merging.

Thanks!

Dirbaio commented:
Feel free to do breaking changes, this isn't even released on crates.io. If the changes improve the API then it's worth it.

Speaking of breaking, with this we'll have two different APIs for the same thing: builder and macro. If the builder can do everything the macro can, then maybe we should remove the macro API
- it's not nice to have two ways of doing the same thing
- the macros have other problems anyway, such as being very hard to learn

alexmoon commented:
Ok, I've updated the macros to use the new builder interface and removed the `register` functions from `GattServer`. I've updated the examples and added a new example showing how to use the builder interface directly. If there are no concerns/objections I'll merge the changes tomorrow.

Dirbaio commented:
I meant removing the *actual* macros. From the user's PoV the macro *is* an API. There's 2 ways of writing the same thing (2 APIs): using the builder and using the macro. 

Unless you guys think there's some value in keeping the macro...

Dirbaio commented:
LGTM then :)
